web/src/test/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateControllerTest.java

```diff
@@ -0,0 +1,313 @@
+
+  private static final String LANGUAGE_SKILL_TOO_LOW_MESSAGE_KEY =
+      "javax.validation.constraints.Min.message";
+
+  private static final String LANGUAGE_SKILL_TOO_HIGH_MESSAGE_KEY =
+      "javax.validation.constraints.Max.message";
+
+  private final String missingName = null;
+  private final String emptyName = StringUtils.EMPTY;
+  private final String tooLongName = StringUtils.repeat('a', 101);
+  private final String correctName = "John Doe";
+
+  private final String missingEmail = null;
+  private final String emptyEmail = StringUtils.EMPTY;
+  private final String tooLongEmail = StringUtils.repeat('a', 400) + "@email.com";
+  private final String incorrectEmail = "malformed";
+  private final String correctEmail = "john@doe.com";
+
+  private final String tooLongPhone = StringUtils.repeat('0', 21);
+  private final String incorrectPhone = "malformed";
+  private final String correctPhone = "06301111111";
+
+  private final String tooLongReferer = StringUtils.repeat('a', 21);
+  private final String correctReferer = "John Doe";
+
+  private final Short tooLowLanguageSkill = -1;
+  private final Short tooHighLanguageSkill = 11;
+  private final Short correctLanguageSkill = 10;
+
+  private final String correctDescription = "correctDescription";
+
+  private final Long candidateId = 1L;
+
+  @Mock
+  private CandidateService candidateService;
+
+  @Mock
+  private MessageSource messageSource;
+
+  @InjectMocks
+  private SingleCandidateController singleCandidateController;
+
+  /**
+   * Contains a correct email and a correct name so there won't be {@code NotNull} validation errors
+   * for these fields. In each method where we expect error response this object gets used and
+   * filled with malformed input. That way we can ensure that only one error message will be
+   * returned per test method.
+   */
+  private CandidateDTO baseCandidateDto;
+
+  /**
+   * All fields are filled with correct values.
+   */
+  private CandidateDTO correctCandidateDto;
+
+  @Override
+  protected Object[] controllersUnderTest() {
+    return new Object[]{singleCandidateController};
+  }
+
+  @Override
+  public void setUp() {
+    objectMapper = new ObjectMapper();
+
+    LocalValidatorFactoryBean validatorFactoryBean = new LocalValidatorFactoryBean();
+
+    validatorFactoryBean.setValidationMessageSource(messageSource);
+
+    // always return the message key so we don't have to mock each call to return
+    // a message key specific error message
+    given(messageSource.getMessage(anyString(), any(Object[].class), any(Locale.class)))
+        .willAnswer(i -> i.getArgumentAt(0, String.class));
+
+    mockMvc =
+        MockMvcBuilders.standaloneSetup(controllersUnderTest())
+            .setViewResolvers(viewResolver())
+            .setControllerAdvice(controllerAdvice())
+            .setHandlerExceptionResolvers(uncheckedExceptionResolver())
+            .setValidator(validatorFactoryBean)
```

Comment:
Validator setup can go into the AbstractControllerTest class, so it can be reused by all controller tests.